paddle/fluid/framework/tvm/tvm_runtime_batch_matmul_test.cu

```diff
@@ -0,0 +1,297 @@
+  T *z = z_tensor.data<T>();
+
+  int M = x_tensor.dims()[x_tensor.dims().size() - 2];
+  int K = x_tensor.dims()[x_tensor.dims().size() - 1];
+  int B = x_tensor.numel() / M / K;
+  int N = y_tensor.dims()[y_tensor.dims().size() - 1];
+
+  for (int b = 0; b < B; ++b) {
+    for (int m = 0; m < M; ++m) {
+      for (int n = 0; n < N; ++n) {
+        T tmp = 0;
+        for (int k = 0; k < K; ++k) {
+          auto idx_x = b * (M * K) + m * K + k;
+          auto idx_y = b * (K * N) + k * N + n;
+          tmp += x[idx_x] * y[idx_y];
+        }
+        auto idx_z = b * (M * N) + m * N + n;
+        z[idx_z] = tmp;
+      }
+    }
+  }
+  return z_tensor;
+}
+
+template <typename T>
+T MaxDiff(const Tensor &tensor1_in, const Tensor &tensor2_in) {
+  Tensor tensor1 = ConvertToCpuTensor<T>(tensor1_in);
+  Tensor tensor2 = ConvertToCpuTensor<T>(tensor2_in);
+  if (tensor1.dims() == tensor2.dims()) {
+    auto *x = tensor1.data<T>();
+    auto *y = tensor2.data<T>();
+    int64_t n = tensor1.numel();
+    T max_diff = 0;
+    for (int64_t i = 0; i < n; ++i) {
+      max_diff = std::max(std::abs(x[i] - y[i]), max_diff);
+    }
+    return max_diff;
+  } else {
+    return static_cast<T>(-1);
+  }
+}
+
+template <typename T>
+bool TestMain(int batch_size, int feature_size, int M, int K, int N,
+              bool trans_x = false, bool trans_y = false,
+              int iterations = 100) {
+  platform::CUDAPlace place(0);
+  platform::CUDADeviceContext ctx(place);
+
+  Tensor x, y;
+  if (trans_x) {
+    x.Resize({batch_size, feature_size, K, M});
+  } else {
+    x.Resize({batch_size, feature_size, M, K});
+  }
+  x.mutable_data<T>(place);
+  FillRandom(x.data<T>(), x.numel(), place);
+
+  if (trans_y) {
+    y.Resize({batch_size, feature_size, N, K});
+  } else {
+    y.Resize({batch_size, feature_size, K, N});
+  }
+  y.mutable_data<T>(place);
+  FillRandom(y.data<T>(), y.numel(), place);
+
+  Tensor z1, z2;
+  Tensor z_actual = GetActualMatMulValue<T>(x, y);
+  {
+    CuBlasBatchMatMul<T> cublas_matmul(ctx, x, y, &z1, trans_x, trans_y);
+    std::cout << "CuBlas Prepared" << std::endl;
+    auto time = Run(ctx, cublas_matmul, iterations);
+    std::cout << "CuBlas Time: " << time / static_cast<double>(iterations)
+              << "us" << std::endl;
+  }
+
+  {
+    std::string lib_name = "tvm_batch_matmul_" + std::to_string(N) + "_" +
+                           std::to_string(K) + ".so";
+    TVMBatchMatMul<T> tvm_matmul(lib_name, "FP32_matmul", ctx, x, y, &z2,
+                                 trans_x, trans_y);
+    std::cout << "TVM Prepared" << std::endl;
+    auto time = Run(ctx, tvm_matmul, iterations);
+    std::cout << "TVM Time: " << time / static_cast<double>(iterations) << "us"
+              << std::endl;
+  }
+
+  std::cout << "Diff1: " << MaxDiff<T>(z1, z_actual) << std::endl;
+  std::cout << "Diff2: " << MaxDiff<T>(z2, z_actual) << std::endl;
+  return true;
+}
+
+TEST(compare_cublas_tvm, matmul_K_17_N_128) {
```

Comment:
maybe tvm is not good at optimizing a single op. It's good at optimizing a full model (a subgraph), fusing many operators (e.g. elementwise).